nntrainer/tensor/quantizer.h

```diff
@@ -0,0 +1,326 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * @file	quantizer.h
+ * @date	10 December 2024
+ * @brief	This defines quantizers for different types of quantization schemes
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Donghyeon Jeong <dhyeon.jeong@samsung.com>
+ * @bug		No known bugs except for NYI items
+ */
+
+#ifndef __QUANTIZER_H__
+#define __QUANTIZER_H__
+#ifdef __cplusplus
+
+#include <tensor.h>
+
+namespace nntrainer {
+
+/**
+ * @brief defines the quantization scheme
+ * @details NNTrainer provides basic quantization schemes (e.g., Per tensor
+ * affine quantization). Various quantization schemes will be continuously
+ * updated. If you would like to use a different quantization technique, please
+ * select a custom quantizer scheme.
+ */
+enum class QScheme : uint8_t {
+  /** predefined quantizer */
+  PER_TENSOR_AFFINE = 0x00,
+  PER_CHANNEL_AFFINE = 0x01,
+  BINARY_CODE_BASED = 0x02,
+  /** this is for custom use */
+  CUSTOM_QUANTIZER_01 = 0x10,
+  CUSTOM_QUANTIZER_02 = 0x11,
+  CUSTOM_QUANTIZER_03 = 0x12,
+  CUSTOM_QUANTIZER_04 = 0x13,
+  CUSTOM_QUANTIZER_05 = 0x14,
+  CUSTOM_QUANTIZER_06 = 0x15,
+};
+
+/**
+ * @class Quantizer class
+ * @brief Quantizer class is a base class for all quantizers.
+ * @note A custom quantizer must inherit this class and implement virtual
+ * functions.
+ */
+class Quantizer {
+private:
+  static inline std::unordered_map<QScheme, Quantizer *>
+    custom_quantizers; /** Hash table that holds empty instances of the custom
+                          quantizers */
+
+protected:
+  /**
+   * @brief Register the user defined quantizer class
+   *
+   * @param qscheme Quantization scheme (use CUSTOM_QUANTIZER_#)
+   * @param quantizer quantizer class to register
+   *
+   * @note This function registers the custom quantizer class. User defined
+   * derived class must be registered with this function.
+   */
+  static void registerQuantizer(QScheme qscheme, Quantizer &quantizer) {
```

Comment:
This needs to check the duplication of the map. If it is static, the only one object exists throughout the process. I just wonder wether it is possible to use same type of tensor can have different quantizer.